src/safemode_ui.cpp

```diff
@@ -0,0 +1,740 @@
+        mvwputch( w_header, 3, pos.second, c_ltgray, LINE_XOXO );
+    }
+
+    mvwprintz( w_header, 3, 1, c_white, "#" );
+    mvwprintz( w_header, 3, column_pos[COLUMN_RULE] + 4, c_white, _( "Rules" ) );
+    mvwprintz( w_header, 3, column_pos[COLUMN_ATTITUDE] + 2, c_white, _( "Attitude" ) );
+    mvwprintz( w_header, 3, column_pos[COLUMN_PROXIMITY] + 2, c_white, _( "Dist" ) );
+    mvwprintz( w_header, 3, column_pos[COLUMN_INCLUDE_EXCLUDE] + 2, c_white, _( "I/E" ) );
+
+    wrefresh( w_header );
+
+    int tab = GLOBAL_TAB;
+    int line = 0;
+    int column = 0;
+    int start_pos = 0;
+    bool changes_made = false;
+    input_context ctxt( "SAFEMODE" );
+    ctxt.register_cardinal();
+    ctxt.register_action( "CONFIRM" );
+    ctxt.register_action( "QUIT" );
+    ctxt.register_action( "NEXT_TAB" );
+    ctxt.register_action( "PREV_TAB" );
+    ctxt.register_action( "ADD_RULE" );
+    ctxt.register_action( "REMOVE_RULE" );
+    ctxt.register_action( "COPY_RULE" );
+    ctxt.register_action( "ENABLE_RULE" );
+    ctxt.register_action( "DISABLE_RULE" );
+    ctxt.register_action( "MOVE_RULE_UP" );
+    ctxt.register_action( "MOVE_RULE_DOWN" );
+    ctxt.register_action( "TEST_RULE" );
+    ctxt.register_action( "HELP_KEYBINDINGS" );
+
+    if( is_safemode_in ) {
+        ctxt.register_action( "SWITCH_SAFEMODE_OPTION" );
+        ctxt.register_action( "SWAP_RULE_GLOBAL_CHAR" );
+    }
+
+    std::ostringstream temp_stream;
+
+    while( true ) {
+        int locx = 17;
+        locx += shortcut_print( w_header, 2, locx, c_white,
+                                ( tab == GLOBAL_TAB ) ? hilite( c_white ) : c_white, _( "[<Global>]" ) ) + 1;
+        shortcut_print( w_header, 2, locx, c_white,
+                        ( tab == CHARACTER_TAB ) ? hilite( c_white ) : c_white, _( "[<Character>]" ) );
+
+        locx = 55;
+        mvwprintz( w_header, 0, locx, c_white, _( "Safemode enabled:" ) );
+        locx += shortcut_print( w_header, 1, locx,
+                                ( ( get_option<bool>( "SAFEMODE" ) ) ? c_ltgreen : c_ltred ), c_white,
+                                ( ( get_option<bool>( "SAFEMODE" ) ) ? _( "True" ) : _( "False" ) ) );
+        locx += shortcut_print( w_header, 1, locx, c_white, c_ltgreen, "  " );
+        locx += shortcut_print( w_header, 1, locx, c_white, c_ltgreen, _( "<S>witch" ) );
+        shortcut_print( w_header, 1, locx, c_white, c_ltgreen, "  " );
+
+        wrefresh( w_header );
+
+        // Clear the lines
+        for( int i = 0; i < content_height; i++ ) {
+            for( int j = 0; j < 79; j++ ) {
+                mvwputch( w, i, j, c_black, ' ' );
+            }
+
+            for( auto &pos : column_pos ) {
+                mvwputch( w, i, pos.second, c_ltgray, LINE_XOXO );
+            }
+        }
+
+        auto &current_tab = ( tab == GLOBAL_TAB ) ? global_rules : character_rules;
+        const bool current_tab_non_empty = !current_tab.empty();
```

Comment:
What's the advantages of using this variable instead of `!current_tab.empty()` directly? The later is always up to date, while this variable can easily be wrong:
```C++
const bool is_non_empty = !foo.empty();
...
foo.clear();
...
if(is_non_empty) {
    foo[0]; // boom
}
```